### ISSUE TITLE
hashFullRequest break the proxy

### ISSUE THREAD

**Issue by bdecarne** [seglo/connect-prism#21]
When i set hashFullRequest to `true` and try to mockrecord a request with a request body, the proxy wait forever.

It seems that the call of `mockFilenameGenerator.getBody(req);` in `proxy.js` break `prism.server.web(req, res);` behaviour.

Any relation with https://github.com/nodejitsu/node-http-proxy/issues/667 ?

**Comment by seglo** [comment]
I recently added `winston` logging to `connect-prism`.  You can enable verbose logging with `useVerboseLog()`

https://github.com/seglo/connect-prism#logging

```javascript
var prism = require('connect-prism');
prism.useVerboseLog();
```

**Comment by seanders** [comment]
@seglo  I'm having a similar issue. The POST body seems to not get send through the proxy server. Any follow up on this issue?

**Comment by fcuenya** [comment]
I think I came up with a fix for this. I really like this project and the hashFullRequest feature was a deal breaker for me, so I spent some time digging through the code to try to make it work.

It works well for my use case so far, but I'll like to check in the project, run the tests and do some more testing before I commit the fix (I'm new to GitHub so please bear with me :-))

In the meantime I'll leave my code here, so that you guys can have a look at it. I only had to change the proxyResponse method from the Proxy class (proxy.js):

```javascript
function proxyResponse(req, res, prism) {
    if (prism.config.hashFullRequest) {
      var passThroughStream = stream.PassThrough();
      req.pipe(passThroughStream);
      prismUtils.getBody(passThroughStream, function(bodyContent) {
        req.body = bodyContent;
      });
    }
    prism.server.web(req, res);
    logger.logSuccess('Proxied', req, prism);
  } 
```

Also declare the stream variable:

```javascript
var stream = require('stream');
```

**Comment by seglo** [comment]
That's interesting.  Could you elaborate on the issue you were experiencing and why this solved it for you?  What was the reasoning to try using PassThrough?

As I've mentioned earlier in this thread I haven't found a way to reproduce this issue.  The code you shared does pass all tests in `master`.  If you're preparing a pull request could you also include a test which demonstrates a failure when not using this code?  Or you could also describe the use case so I could write a test myself.

I know it's been awhile since this issue was opened, but if any of the other commenters could try the @fcuenya fix I would appreciate it.

**Comment by fcuenya** [comment]
Sure, so here's what I think it's happening: 

- When using hashFullRequest the request gets read to retrieve the body before it gets passed to the proxy server. On my example I was using a POST request with user+password on the body (I know, I know... :-))

- Since the request is a stream, the step above would consume and end it, so when the proxy server tried to read from it, it would hang there waiting for data to come. Because of, this the request never got sent to the "real" server, timing out (eventually).

What I did by adding the PassThrough stream was "sort of multiplexing" this request stream, so that its data could be read by the getBody method without ending it (just ending the passthrough stream instead). Once the body content had been read, the callback function would set the req.body parameter with such content so that it can be used through the rest of the process.

I was already planning on include some testing to demonstrate all these with the pull request, I just need to find some time to get my repo setup and work on that :-)

So far this "cool thingy" you guys created is working well for mocking out the backend in our project, and hopefully we can leverage it for TDD. Good job!!

**Comment by seglo** [comment]
I see.  Thanks for the explanation.  I always thought that I was just listening for events on a stream for the body parsing... so when the proxy was reading it my own events would fire and get the same data before/after the proxy was.  But you're saying that once the stream is read it's not available to other handlers reading it?

**Comment by fcuenya** [comment]
That's my understanding of how streams work, sort of like a buffer in this case. Once you read the data, via the on('data') event, that data is no longer in the stream and other would replace it. Since you read all the data in the getBody method, you are effectively consuming the whole stream before the actual proxy can send it to the real server.

Piping it with a passthrough gives you two independent streams to read the same data from. This way you can close one (the passthrough) and still have your data available to read from the other (the actual request).

**Comment by fcuenya** [comment]
I just created some tests for this fix and I will be checking these changes shortly (once I figure out how this whole github process work... getting there!)

**Comment by seglo** [comment]
Thanks for the PR @fcuenya.  I've released your patch with `1.0.4`.

**Comment by fcuenya** [comment]
Awesome! Thank you for this cool piece of software!
